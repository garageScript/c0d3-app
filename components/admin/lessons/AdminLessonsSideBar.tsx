import React from 'react'
import _ from 'lodash'
import { Lesson } from '../../../graphql/index'

type SideBarLessonProps = {
  lessons: Lesson[] | undefined
  selectedLesson: number
  setLessons: React.Dispatch<React.SetStateAction<Lesson[] | null>>
  setSelectedLesson: React.Dispatch<React.SetStateAction<number>>
}

export const AdminLessonsSideBar: React.FC<SideBarLessonProps> = ({
  lessons,
  setSelectedLesson,
  selectedLesson
}) => {
  const lessonListData = _.cloneDeep(lessons) || []
  if (lessonListData.length === 0) {
    lessonListData.push({
      id: 0,
      title: 'Create New Lesson',
      description: '',
      order: -1,
      challenges: []
    })
  }
  //check if create new lesson has already been pushed to the lessons array
  const lastIndex = lessonListData.length - 1
  const { title } = lessonListData[lastIndex]
  if (title !== 'Create New Lesson') {
    lessonListData.push({
      id: 0,
      title: 'Create New Lesson',
      description: '',
      order: -1,
      challenges: []
    })
  }

  const lessonList = lessonListData.map((obj: any, arrIndex: number) => (
    <a
      key={arrIndex}
      data-testid="challenge-title"
      role="tab"
      className={'nav-link' + (arrIndex === selectedLesson ? ' active' : '')}
      data-toggle="pill"
      style={{ wordBreak: 'break-word', cursor: 'pointer' }}
      onClick={() => setSelectedLesson(arrIndex)}
    >
      {obj.title}
    </a>
  ))

  return (
    <div className="col-4 mt-4 ">
      <div
        className="nav flex-column nav-pills"
        role="tablist"
        aria-orientation="vertical"
      >
        {lessonList}
      </div>
    </div>
  )
}
