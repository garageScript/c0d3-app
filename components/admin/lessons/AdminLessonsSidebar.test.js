import * as React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { AdminLessonsSideBar } from './AdminLessonsSideBar'
import dummyLessonData from '../../../__dummy__/lessonData'
describe('AdminLessonSideBar component', () => {
  test('Should render lessons', async () => {
    const { container } = render(
      <AdminLessonsSideBar
        lessons={dummyLessonData}
        selectedLesson={1}
        setLessons={() => {}}
        setSelectedLesson={() => {}}
      />
    )
    expect(container).toMatchSnapshot()
    const lesson = screen.getByRole('tab', { name: 'Arrays' })
    await userEvent.click(lesson)
    expect(container).toMatchSnapshot()
  })
  test('Should render empty lessons', async () => {
    const { container } = render(
      <AdminLessonsSideBar
        selectedLesson={1}
        setLessons={() => {}}
        setSelectedLesson={() => {}}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
