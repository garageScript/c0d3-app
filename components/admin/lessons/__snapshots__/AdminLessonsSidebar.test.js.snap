// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminLessonSideBar component Should render empty lessons 1`] = `
<div>
  <div
    class="col-4 mt-4 "
  >
    <div
      aria-orientation="vertical"
      class="nav flex-column nav-pills"
      role="tablist"
    >
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Create New Lesson
      </a>
    </div>
  </div>
</div>
`;

exports[`AdminLessonSideBar component Should render lessons 1`] = `
<div>
  <div
    class="col-4 mt-4 "
  >
    <div
      aria-orientation="vertical"
      class="nav flex-column nav-pills"
      role="tablist"
    >
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Foundations of JavaScript
      </a>
      <a
        class="nav-link active"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Variables & Functions
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Arrays
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Objects
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Front End Engineering
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        End To End
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        React, GraphQL, SocketIO
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        JavaScript Algorithms
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Trees
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        General Algorithms
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Create New Lesson
      </a>
    </div>
  </div>
</div>
`;

exports[`AdminLessonSideBar component Should render lessons 2`] = `
<div>
  <div
    class="col-4 mt-4 "
  >
    <div
      aria-orientation="vertical"
      class="nav flex-column nav-pills"
      role="tablist"
    >
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Foundations of JavaScript
      </a>
      <a
        class="nav-link active"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Variables & Functions
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Arrays
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Objects
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Front End Engineering
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        End To End
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        React, GraphQL, SocketIO
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        JavaScript Algorithms
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Trees
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        General Algorithms
      </a>
      <a
        class="nav-link"
        data-testid="challenge-title"
        data-toggle="pill"
        role="tab"
        style="word-break: break-word; cursor: pointer;"
      >
        Create New Lesson
      </a>
    </div>
  </div>
</div>
`;
